File: typescript/packages/node/src/core/nextcloud/rm.ts
import { invalidateSubtree } from '@struktoai/mirage-core/cache/context'
import type { PathSpec } from '@struktoai/mirage-core/types'
import { enoent } from '@struktoai/mirage-core/utils/errors'
import { rstripSlash } from '@struktoai/mirage-core/utils/slash'
import type { NextcloudAccessor } from '../../accessor/nextcloud.ts'
import { isNotFound, nextcloudKey } from './util.ts'

export async function rmR(accessor: NextcloudAccessor, path: PathSpec): Promise<void> {
  const key = `${rstripSlash(nextcloudKey(path))}/`
  const op = await accessor.operator()
  try {
    await op.removeAll(key)
  } catch (error) {
    if (isNotFound(error)) throw enoent(path)
    throw error
  }
  await invalidateSubtree(path)
}

File: python/mirage/core/nextcloud/rm.py
from opendal.exceptions import NotFound

from mirage.accessor.nextcloud import NextcloudAccessor
from mirage.cache.context import invalidate_subtree
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def rm_r(accessor: NextcloudAccessor, path: PathSpec) -> None:
    raw = path.mount_path
    key = raw.strip("/") + "/"
    op = accessor.operator()
    try:
        await op.remove_all(key)
    except NotFound as exc:
        raise enoent(path) from exc
    await invalidate_subtree(path)
